src/events/controller.py

```diff
@@ -0,0 +1,215 @@
+                (username, self.charm.workload.generate_password()) for username in INTERNAL_USERS
+            ]
+            for username, password in credentials:
+                self.charm.state.cluster.update({f"{username}-password": password})
+
+        # cluster-uuid is only created on the broker (`cluster-manager` in large deployments)
+        if not self.charm.state.cluster.cluster_uuid and self.charm.state.runs_broker:
+            uuid = self.controller_manager.generate_uuid()
+            self.charm.state.cluster.update({"cluster-uuid": uuid})
+            self.charm.state.peer_cluster.update({"cluster-uuid": uuid})
+
+        # Controller is tasked with populating quorum bootstrap config
+        if self.charm.state.runs_controller and not self.charm.state.cluster.bootstrap_controller:
+            quorum_uris = {"controller-quorum-uris": self.charm.state.controller_quorum_uris}
+            self.charm.state.cluster.update(quorum_uris)
+
+            generated_password = self.charm.workload.generate_password()
+
+            if self.charm.state.peer_cluster_orchestrator:
+                self.charm.state.peer_cluster_orchestrator.update(quorum_uris)
+
+                if not self.charm.state.peer_cluster_orchestrator.controller_password:
+                    self.charm.state.peer_cluster_orchestrator.update(
+                        {"controller-password": generated_password}
+                    )
+            elif not self.charm.state.peer_cluster.controller_password:
+                # single mode, controller & leader
+                self.charm.state.cluster.update({"controller-password": generated_password})
+
+            bootstrap_data = {
+                "bootstrap-controller": self.charm.state.bootstrap_controller,
+                "bootstrap-unit-id": str(self.charm.state.kraft_unit_id),
+                "bootstrap-replica-id": self.controller_manager.generate_uuid(),
+            }
+            self.charm.state.cluster.update(bootstrap_data)
+
+    def _format_storages(self) -> None:
+        """Format storages provided relevant keys exist."""
+        if not self.charm.state.kraft_mode:
+            return
+
+        self.broker.config_manager.set_server_properties()
+        if self.charm.state.runs_broker:
+            credentials = self.charm.state.cluster.internal_user_credentials
+        elif self.charm.state.runs_controller:
+            credentials = {
+                self.charm.state.peer_cluster.broker_username: self.charm.state.peer_cluster.broker_password
+            }
+
+        self.controller_manager.format_storages(
+            uuid=self.charm.state.peer_cluster.cluster_uuid,
+            internal_user_credentials=credentials,
+            initial_controllers=f"{self.charm.state.peer_cluster.bootstrap_unit_id}@{self.charm.state.peer_cluster.bootstrap_controller}:{self.charm.state.peer_cluster.bootstrap_replica_id}",
+        )
+
+    def _leader_elected(self, event: LeaderElectedEvent) -> None:
```

Comment:
**todo**: I'm a bit itchy when seeing/using `leader_elected` event. Following up @marcoppenheimer [comment](https://github.com/canonical/kafka-operator/pull/281#discussion_r1889413325) I would rather have this logic happen in `update_status` or `config_changed` (since update status calls config_changed). 
- afaik, there are no guarantees that `leader_elected` triggers _ever_ after the fist deployment
- Triggers of `leader_elected` tend to happen in the context of controller failing to see a unit, which can mean unit failure, so this code being close to failure states is not the best :)